docs: Add documentation for the io.IOBase class.

andrewleech opened the pull request:
### Summary

Documents the `io.IOBase` class which is used to create custom stream objects in Python. Subclasses override `readinto`, `write`, and `ioctl` to create objects that work with `print()`, `json.dump()`, `select.poll()`, user filesystems, etc.

The docs focus on the three methods the C stream layer actually dispatches to, rather than documenting the full stream API which is provided automatically by the C infrastructure. The ioctl section covers the common operations (flush, poll, close, buffer size) in the same style as vfs.AbstractBlockDev.ioctl.

Also fixes a statement in io.rst that incorrectly claimed streams cannot be subclassed in pure Python.

### Testing

Built docs locally with `make html`, renders correctly.

### Generative AI

I used generative AI tools when creating this PR, but a human has checked the code and is responsible for the description above.

pi-anl committed:
docs/library/io: Add documentation for the io.IOBase class.

Document the three methods that IOBase subclasses implement (readinto,
write, ioctl) and the common ioctl operations. Includes a simple write-only
stream example and a pollable ring buffer example.

IOBase docs are inline in io.rst before StringIO/BytesIO, following
CPython's structure.

Also fix a statement in io.rst that incorrectly claimed streams cannot be
subclassed in pure Python.

Signed-off-by: Andrew Leech <andrew.leech@planet-innovation.com>

Josverl commented:
Thanks Andrew, I think this a very good addition.
I had been trying to add version of this to the micropython-stubs project earlier, but did not get the quality i was after.
.

andrewleech commented:
Thanks for the feedback, I've rewritten to address suggestions and improve the overall writing style.

- Renamed `ioctl(request, arg=0)` to `ioctl(op, arg)` to match vfs.AbstractBlockDev. All ioctl methods across the codebase are positional-only so the parameter names are cosmetic, but they should be consistent in docs regardless.
- Replaced the 230-line example with two shorter ones: a minimal write-only stream (~15 lines) and a pollable ring buffer with select.poll() (~35 lines). Both tested on the unix port.
- Cut the methods section down to just the three methods that IOBase subclasses actually implement: `readinto(buf)`, `write(buf)`, `ioctl(op, arg)`. The previous version documented read(), readline(), seek(), tell(), flush(), close() which are all provided automatically by the C stream infrastructure and don't need to be implemented by subclasses.
- Trimmed the ioctl ops list to the ones Python subclasses commonly handle (flush, poll, close, buffer size) in the same terse numbered-list style as vfs.rst. The rarely-used ops (seek, timeout, fileno etc.) get a brief mention pointing to py/stream.h.
- Added a CPython difference admonition since MicroPython's IOBase is much more minimal.
- Fixed a statement in io.rst that incorrectly claimed streams can't be subclassed in pure Python.

dpgeorge commented on docs/library/io.rst:
```diff
@@ -119,3 +187,75 @@ Classes
+                return 0
+            if op == _MP_STREAM_CLOSE:
+                return 0
+            return 0
```
This should be `return -1` to indicate an unsupported op.
